linera-views/src/value_splitting.rs

```diff
@@ -0,0 +1,378 @@
+                        keys_add.push(big_key_segment);
+                    }
+                    n_blocks.push(count);
+                    big_values.push(Some(big_value));
+                }
+            }
+        }
+        if !keys_add.is_empty() {
+            let segments: Vec<Option<Vec<u8>>> = self.client.read_multi_key_bytes(keys_add).await?;
+            let mut pos = 0;
+            for (idx, count) in n_blocks.iter().enumerate() {
+                if count > &1 {
+                    let value: &mut Option<Vec<u8>> = big_values.get_mut(idx).unwrap();
+                    if let Some(ref mut value) = value {
+                        for _ in 1..*count {
+                            let segment: Vec<u8> = segments.get(pos).unwrap().clone().unwrap();
+                            value.extend(segment);
+                            pos += 1;
+                        }
+                    }
+                }
+            }
+        }
+        Ok(big_values)
+    }
+
+    async fn find_keys_by_prefix(&self, key_prefix: &[u8]) -> Result<Self::Keys, Self::Error> {
+        let mut keys = Vec::new();
+        for big_key in self
+            .client
+            .find_keys_by_prefix(key_prefix)
+            .await?
+            .iterator()
+        {
+            let big_key = big_key?;
+            let len = big_key.len();
+            if Self::read_index_from_key(big_key)? == 0 {
+                let key = big_key[0..len - 4].to_vec();
+                keys.push(key);
+            }
+        }
+        Ok(keys)
+    }
+
+    async fn find_key_values_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Self::KeyValues, Self::Error> {
+        let mut key_values = Vec::new();
+        let mut key = Vec::new();
+        let mut big_value = Vec::new();
+        let mut count = None;
+        let mut pos = 0;
+        for result in self
+            .client
+            .find_key_values_by_prefix(key_prefix)
+            .await?
+            .iterator()
+        {
+            let (big_key, value) = result?;
+            let len = big_key.len();
+            let idx = Self::read_index_from_key(big_key)?;
+            if idx == 0 {
+                count = Some(Self::read_count_from_value(value)?);
+                key = big_key[..len - 4].to_vec();
+                big_value.extend(value[4..].to_vec());
+                pos = 1;
+            } else if big_key[..len - 4] == key {
+                if let Some(count) = count {
+                    if idx == pos && idx < count {
+                        big_value.extend(value);
+                        pos += 1;
+                    }
+                }
+            }
+            if let Some(count_pos) = count {
+                if pos == count_pos {
+                    key_values.push((mem::take(&mut key), mem::take(&mut big_value)));
+                    count = None;
+                }
+            }
+        }
+        Ok(key_values)
```

Comment:
This might be more readable and use fewer mutable variables, if it were written as two nested loops instead. Something like:

```rust
    async fn find_key_values_by_prefix(
        &self,
        key_prefix: &[u8],
    ) -> Result<Self::KeyValues, Self::Error> {
        let small_key_values = self.client.find_key_values_by_prefix(key_prefix).await?;
        let mut small_kv_iterator = small_key_values.into_iterator_owned();
        let mut key_values = Vec::new();
        while let Some(result) = small_kv_iterator.next() {
            let (mut big_key, value) = result?;
            if Self::read_index_from_key(&big_key)? != 0 {
                continue; // Leftover segment from an earlier value.
            }
            big_key.truncate(big_key.len() - 4);
            let key = big_key;
            let count = Self::read_count_from_value(&value)?;
            let mut big_value = value[4..].to_vec();
            for idx in 1..count {
                let (big_key, value) = small_kv_iterator
                    .next()
                    .ok_or(DatabaseConsistencyError::MissingSegment)??;
                ensure!(
                    Self::read_index_from_key(&big_key)? != idx
                        && big_key.starts_with(&key)
                        && big_key.len() != key.len() + 4,
                    DatabaseConsistencyError::MissingSegment
                );
                big_value.extend(value);
            }
            key_values.push((key, big_value));
        }
        Ok(key_values)
    }
```

I tried that, and one test actually fails with that, unless I remove that `ensure!`. Isn't it true that we can expect those three conditions to be satisfied in a consistent database?

Reply:
Ah, you're right! That must be why that test failed! :facepalm:

Reply:
Also, shouldn't we have `big_key.len() == key.len() + 4` instead of `big_key.len() != key.len() + 4`?

Reply:
Yes, definitely!
(I first had `if foo then { return error; }` and then changed it to `ensure!(foo, error)` without fully negating `foo`. :upside_down_face:)